/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2023-06-07T14:55:23.017
 */
package bio.ferlab.etl.model

import java.sql.Timestamp


case class NormalizedConsequences(`chromosome`: String = "1",
                                  `start`: Long = 14464,
                                  `reference`: String = "A",
                                  `alternate`: String = "T",
                                  `name`: Option[String] = None,
                                  `end`: Long = 14465,
                                  `consequences`: Seq[String] = Seq("non_coding_transcript_exon_variant"),
                                  `impact`: String = "MODIFIER",
                                  `symbol`: String = "WASH7P",
                                  `ensembl_gene_id`: String = "ENSG00000227232",
                                  `ensembl_feature_id`: String = "ENST00000488147",
                                  `ensembl_transcript_id`: String = "ENST00000488147",
                                  `ensembl_regulatory_id`: Option[String] = None,
                                  `feature_type`: String = "Transcript",
                                  `strand`: String = "-1",
                                  `biotype`: String = "unprocessed_pseudogene",
                                  `variant_class`: String = "SNV",
                                  `exon`: Exon = Exon(),
                                  `intron`: Intron = Intron(),
                                  `hgvsc`: String = "ENST00000488147.1:n.1291T>A",
                                  `hgvsp`: Option[String] = None,
                                  `hgvsg`: String = "chr1:g.14464A>T",
                                  `cds_position`: Option[String] = None,
                                  `cdna_position`: String = "1291",
                                  `protein_position`: Option[String] = None,
                                  `amino_acids`: AMINO_ACIDS = AMINO_ACIDS(),
                                  `codons`: CODONS = CODONS(),
                                  `original_canonical`: Boolean = true,
                                  `refseq_mrna_id`: Option[Seq[String]] = None,
                                  `aa_change`: Option[String] = None,
                                  `coding_dna_change`: Option[String] = None,
                                  `impact_score`: Int = 1)

