/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2023-06-07T14:46:01.610
 */
package bio.ferlab.etl.model

case class VCF_SNV_INPUT(`contigName`: String = "chr1",
                         `start`: Long = 14463,
                         `end`: Long = 14464,
                         `names`: Option[Seq[String]] = None,
                         `referenceAllele`: String = "A",
                         `alternateAlleles`: Seq[String] = Seq("T"),
                         `qual`: Double = 1387.08,
                         `INFO_FILTERS`: Seq[String] = Seq("DRAGENSnpHardQUAL"),
                         `splitFromMultiAllelic`: Boolean = false,
                         `INFO_NEGATIVE_TRAIN_SITE`: Boolean = true,
                         `INFO_AC`: Seq[Int] = Seq(18),
                         `INFO_culprit`: String = "MQRankSum",
                         `INFO_SOR`: Double = 1.688,
                         `INFO_ReadPosRankSum`: Double = 0.967,
                         `INFO_AN`: Int = 124,
                         `INFO_InbreedingCoeff`: Double = 0.1891,
                         `INFO_AF`: Seq[Double] = Seq(0.145),
                         `INFO_FS`: Double = 0.0,
                         `INFO_DP`: Int = 202,
                         `INFO_POSITIVE_TRAIN_SITE`: Boolean = false,
                         `INFO_VQSLOD`: Double = -2.042,
                         `INFO_ClippingRankSum`: Double = 0.0,
                         `INFO_RAW_MQ`: Option[Double] = None,
                         `INFO_BaseQRankSum`: Double = 0.967,
                         `INFO_MLEAF`: Seq[Double] = Seq(0.323),
                         `INFO_MLEAC`: Seq[Int] = Seq(40),
                         `INFO_MQ`: Double = 29.26,
                         `INFO_QD`: Double = 27.2,
                         `INFO_END`: Option[Int] = None,
                         `INFO_CSQ`: Seq[INFO_csq] = Seq(INFO_csq()),
                         `INFO_HaplotypeScore`: Option[Double] = None,
                         `INFO_MQRankSum`: Double = 0.967,
                         `INFO_ExcessHet`: Double = 2.0E-4,
                         `INFO_DS`: Boolean = false,
                         `INFO_OLD_MULTIALLELIC`: Option[String] = None,
                         `genotypes`: Seq[GENOTYPES] = Seq(GENOTYPES()),
                         `normalizationStatus`: NORMALIZATIONSTATUS = NORMALIZATIONSTATUS())

case class INFO_csq(`Allele`: String = "T",
                    `Consequence`: Seq[String] = Seq("non_coding_transcript_exon_variant"),
                    `IMPACT`: String = "MODIFIER",
                    `SYMBOL`: String = "WASH7P",
                    `Feature_type`: String = "Transcript",
                    `Gene`: String = "ENSG00000227232",
                    `PUBMED`: Option[String] = None,
                    `PICK`: Option[String] = None,
                    `Feature`: String = "ENST00000488147",
                    `EXON`: Exon = Exon(),
                    `BIOTYPE`: String = "unprocessed_pseudogene",
                    `INTRON`: Intron = Intron(),
                    `HGVSc`: String = "ENST00000488147.1:n.1291T>A",
                    `HGVSp`: Option[String] = None,
                    `STRAND`: String = "-1",
                    `CDS_position`: Option[String] = None,
                    `cDNA_position`: String = "1291",
                    `Protein_position`: Option[String] = None,
                    `Amino_acids`: AMINO_ACIDS = AMINO_ACIDS(),
                    `Codons`: CODONS = CODONS(),
                    `VARIANT_CLASS`: String = "SNV",
                    `HGVSg`: String = "chr1:g.14464A>T",
                    `CANONICAL`: String = "YES",
                    `RefSeq`: Option[String] = None)

case class GENOTYPES(`sampleId`: String = "S11930",
                     `conditionalQuality`: Int = 33,
                     `SB`: Option[Seq[Int]] = None,
                     `alleleDepths`: Seq[Int] = Seq(1, 3),
                     `PID`: String = "14522_G_A",
                     `phased`: Boolean = false,
                     `calls`: Seq[Int] = Seq(0, 1),
                     `MIN_DP`: Option[Int] = None,
                     `PGT`: String = "0|1",
                     `phredLikelihoods`: Seq[Int] = Seq(123, 0, 33),
                     `depth`: Int = 4,
                     `RGQ`: Option[Int] = None)

case class AMINO_ACIDS(`reference`: Option[String] = None,
                       `variant`: Option[String] = None)

case class Exon(`rank`: Option[String] = None,
                `total`: Option[String] = None)

case class NORMALIZATIONSTATUS(`changed`: Boolean = false,
                               `errorMessage`: Option[String] = None)

case class Intron(`rank`: Option[String] = None,
                  `total`: Option[String] = None)

case class CODONS(`reference`: Option[String] = None,
                  `variant`: Option[String] = None)
