/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2023-06-07T14:46:01.610
 */
package bio.ferlab.etl.normalize.model

case class VCF_SNV_INPUT(`contigName`: String = "chr1",
                    `start`: Long = 10229,
                    `end`: Long = 10231,
                    `names`: Option[Seq[String]] = None,
                    `referenceAllele`: String = "AC",
                    `alternateAlleles`: Seq[String] = Seq("A"),
                    `qual`: Double = 65.93,
                    `INFO_FILTERS`: Option[Seq[String]] = None,
                    `splitFromMultiAllelic`: Boolean = false,
                    `INFO_AC`: Seq[Int] = Seq(2),
                    `INFO_SOR`: Double = 0.122,
                    `INFO_ReadPosRankSum`: Double = 0.303,
                    `INFO_AN`: Int = 2,
                    `INFO_InbreedingCoeff`: Option[Double] = None,
                    `INFO_AF`: Seq[Double] = Seq(1.0),
                    `INFO_FS`: Double = 2.869,
                    `INFO_DP`: Int = 68,
                    `INFO_BaseQRankSum`: Option[Double] = None,
                    `INFO_MLEAF`: Seq[Double] = Seq(1.0),
                    `INFO_MLEAC`: Seq[Int] = Seq(2),
                    `INFO_MQ`: Double = 25.07,
                    `INFO_QD`: Double = 4.12,
                    `INFO_END`: Option[Int] = None,
                    `INFO_CSQ`: Seq[INFO_csq] = Seq(INFO_csq()),
                    `INFO_SOMATIC`: Boolean = false,
                    `INFO_DB`: Boolean = false,
                    `INFO_MQRankSum`: Double = 1.15,
                    `INFO_FractionInformativeReads`: Double = 0.235,
                    `INFO_ExcessHet`: Double = 0.0,
                    `INFO_R2_5P_bias`: Double = 0.0,
                    `INFO_OLD_MULTIALLELIC`: Option[String] = None,
                    `genotypes`: Seq[GENOTYPES] = Seq(GENOTYPES()),
                    `normalizationStatus`: NORMALIZATIONSTATUS = NORMALIZATIONSTATUS())

case class INFO_csq(`Allele`: String = "-",
                    `Consequence`: Seq[String] = Seq("upstream_gene_variant"),
                    `IMPACT`: String = "MODIFIER",
                    `SYMBOL`: String = "DDX11L2",
                    `Feature_type`: String = "Transcript",
                    `Gene`: String = "ENSG00000290825",
                    `PUBMED`: Option[String] = None,
                    `PICK`: String = "1",
                    `Feature`: String = "ENST00000456328",
                    `EXON`: Exon = Exon(),
                    `BIOTYPE`: String = "lncRNA",
                    `INTRON`: Intron = Intron(),
                    `HGVSc`: Option[String] = None,
                    `HGVSp`: Option[String] = None,
                    `STRAND`: String = "1",
                    `CDS_position`: Option[String] = None,
                    `cDNA_position`: Option[String] = None,
                    `Protein_position`: Option[String] = None,
                    `Amino_acids`: AMINO_ACIDS = AMINO_ACIDS(),
                    `Codons`: CODONS = CODONS(),
                    `VARIANT_CLASS`: String = "deletion",
                    `HGVSg`: String = "chr1:g.10234del",
                    `CANONICAL`: String = "YES",
                    `RefSeq`: Option[String] = None)

case class GENOTYPES(`sampleId`: String = "S03329",
                     `conditionalQuality`: Int = 18,
//                     `F1R2`: Seq[Int] = Seq(0, 8),
//                     `SQ`: Option[Seq[Double]] = None,
//                     `PRI`: Seq[Double] = Seq(0.0, 29.0, 32.0, 34.77, 63.77, 37.77),
//                     `posteriorProbabilities`: Option[Seq[Double]] = None,
                     `SB`: Option[Seq[Int]] = None,
//                     `F2R1`: Seq[Int] = Seq(4, 4),
                     `alleleDepths`: Seq[Int] = Seq(4, 12),
//                     `ICNT`: Seq[Int] = Seq(16, 14),
//                     `AF`: Seq[Double] = Seq(0.75),
                     `phased`: Boolean = false,
                     `calls`: Seq[Int] = Seq(1, 1),
                     `MIN_DP`: Option[Int] = None,
                     `phredLikelihoods`: Seq[Int] = Seq(80, 18, 0),
                     `depth`: Int = 16,
                     `PGT`: Int = 16,
                     `RGQ`: Option[Int] = None
//                     `SPL`: Seq[Int] = Seq(255, 0, 255),
//                     `PS`: Option[Int] = None,
//                     `MB`: Seq[Int] = Seq(0, 4, 10, 2)
                     )

case class AMINO_ACIDS(`reference`: Option[String] = None,
                       `variant`: Option[String] = None)

case class Exon(`rank`: Option[String] = None,
                `total`: Option[String] = None)

case class NORMALIZATIONSTATUS(`changed`: Boolean = false,
                               `errorMessage`: Option[String] = None)

case class Intron(`rank`: Option[String] = None,
                  `total`: Option[String] = None)

case class CODONS(`reference`: Option[String] = None,
                  `variant`: Option[String] = None)
